keps/sig-node/4205-psi-metric/README.md

```diff
@@ -0,0 +1,646 @@
+Phase 2: includes goal 3
+
+### Non-Goals
+
+* Invest in more opportunities to further use PSI metric for pod evictions,
+userspace OOM kills, and so on, for future KEPs.
+
+## Proposal
+
+### User Stories (Optional)
+
+#### Story 1
+
+Today, to identity disruptions caused by resource crunches, Kubernetes users need to
+install node exporter to read PSI metric. With the feature proposed in this enhancement, 
+PSI metric will be available for customers in the Kubernetes metrics API.
+
+#### Story 2
+
+Kubernetes users want to prevent new pods to be scheduled on the nodes that have resource starvation. By using PSI metric, the kubelet will set Node Condition to avoid pods being scheduled on nodes under high resource pressure. The node controller could then set a (taint on the node based on these new Node Conditions)[https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/#taint-nodes-by-condition].
+
+### Risks and Mitigations
+
+There are no significant risks associated with Phase 1 implementation that involves integrating
+the PSI metric in kubelet from runc, and exposing these through the metric endpoint.
+
+Phase 2 involves utilizing the PSI metric to report node conditions. There is a potential
+risk of early reporting for nodes under pressure. We intend to address this concern
+by conducting careful experimentation with PSI threshold values to identify the optimal
+default threshold to be used for reporting the nodes under heavy resource pressure.
+
+## Design Details
+
+#### Phase 1
+1. Add new Data structures PSIData and PSIStats corresponding to the PSI metric output format as following:
+some avg10=0.00 avg60=0.00 avg300=0.00 total=0
+full avg10=0.00 avg60=0.00 avg300=0.00 total=0
+
+```go
+type PSIData struct {
+	Avg10 *float64 `json:”avg10”`
+	Avg60 *float64 `json:”avg60”`
+Avg300 *float64 `json:”avg300”`
+Total *float64 `json:”total”`
+}
+
+type PSIStats struct {
+	Some *PSIData  `json:”some,omitempty”`
+Full *PSIData  `json:”full,omitempty”`
+}
+```
+
+2. Summary API includes stats for both system and kubepods level cgroups. Extend the Summary API to include PSI metric data for each resource obtained from cadvisor. 
+Note: if cadvisor-less is implemented prior to the implementation of this enhancement, the PSI
+metric data will be available through CRI instead.
+
+##### CPU
+```go
+type CPUStats struct { 
+// PSI stats of the overall node
+	PSI cadvisorapi.PSIStats `json:”psi,omitempty”`
+}
+```
+
+##### Memory
+```go
+type MemoryStats struct {
+	// PSI stats of the overall node
+	PSI cadvisorapi.PSIStats `json:”psi,omitempty”`
+}
+```
+
+##### IO
+```go
+// IOStats contains data about IO usage.
+type IOStats struct {
+	// The time at which these stats were updated.
+	Time metav1.Time  `json:”time”`
+
+ // PSI stats of the overall node
+	PSI cadvisorapi.PSIStats  `json:”psi,omitempty”`
+}
+
+type NodeStats struct {
+	 // Stats about the IO pressure of the node
+	IO *IOStats `json:”io,omitempty”`
```

Comment:
why not cpu or memory? or are they already there and this is only explaining additions?

Reply:
yeah this was about additions only. But added CPU and memory also as it is confusing otherwise.